src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/PooledObjects/Extensions.cs

```diff
@@ -5,12 +5,25 @@
 using System.Collections.Immutable;
 using System.Diagnostics;
 using System.Text;
+using System.Threading;
 using Microsoft.Extensions.ObjectPool;
 
 namespace Microsoft.AspNetCore.Razor.PooledObjects;
 
-internal static partial class PooledObjectExtensions
+internal static partial class Extensions
 {
+    /// <summary>
+    ///  Clears the given reference and returns its instance to the pool.
+    /// </summary>
+    public static void ReturnAndClear<T>(this ObjectPool<T> pool, ref T? obj)
+        where T : class
+    {
+        if (Interlocked.Exchange(ref obj, null) is { } o)
```

Comment:
This feels dangerous. The collection type we're using here is not threadsafe, so the `Interlocked.Exchange` is unnecessary, and the presence of it feels like we're trying to guarantee something that we're not capable of guaranteeing. If this were called across multiple threads, we'd exchange `null` for `null` on one thread and return that to the pool, which won't work (silently erroring at best, I honestly have no idea what the real behavior would be). It feels like we either need to make this method actually threadsafe (and use `Interlocked.CompareExchange`), or we should not use `Interlocked` at all.

Reply:
@333fred - Maybe I'm missing something, but I'm not seeing how this isn't thread-safe. It's true that a thread might exchange `null` for `null`. However, if that happens, `o` would be `null` and wouldn't get returned to the pool. Again, I might be missing something, but I don't think `Interlocked.CompareExchange` could be used here because we'd have to know the value of `obj` in order to pass it as the comparand.

Reply:
The not threadsafe part is PooledArrayBuilder itself. This method is threadsafe, in that nothing will be missed returning to the pool, and not returning the same thing twice, but is not threadsafe in that it will possibly return `null` or some in-progress builder (if another thread has populated the `_builder` field and started filling it). In general, this code is either:

1. A perf deopt, because PooledArrayBuilder isn't threadsafe anyways, so why bother with the expensive Interlocked call, or
2. An indication that PooledArrayBuilder has violated a contract of threadsafety.

I'm very much assuming 2 is not the case, so I just don't see the point in the expensive (and potentially confusing due to implying 2) call to Interlocked usage here.

Reply:
OK. That makes sense -- thanks! I was looking at the wrong thing.